components/eamxx/src/share/field/field_utils.hpp

```diff
@@ -179,6 +179,99 @@ void horiz_contraction(const Field &f_out, const Field &f_in,
   impl::horiz_contraction<ST>(f_out, f_in, weight, comm);
 }
 
+// Utility to compute the contraction of a field along its level dimension.
+// This is equivalent to f_out = einsum('...k->...', weight, f_in).
+// The impl is such that:
+// - f_out, f_in, and weight must be provided and allocated
+// - The last dimension is for the levels (LEV)
+// - There can be only up to 3 dimensions of f_in
+// - Weight is assumed to be (in order of checking/impl):
+//   - rank-1, with only LEV dimension
+//   - rank-2, with only COL and LEV dimensions
+template <typename ST>
+void vert_contraction(const Field &f_out, const Field &f_in,
+                      const Field &weight, const ekat::Comm *comm = nullptr) {
+  using namespace ShortFieldTagsNames;
+
+  const auto &l_out = f_out.get_header().get_identifier().get_layout();
+
+  const auto &l_in = f_in.get_header().get_identifier().get_layout();
+
+  const auto &l_w = weight.get_header().get_identifier().get_layout();
+
+  // Sanity checks before handing off to the implementation
+  EKAT_REQUIRE_MSG(
+      l_w.rank() >= 1 && l_w.rank() <= 2,
+      "Error! The weight field must be at least rank-1 and at most rank-2.\n"
+      "The weight field has rank "
+          << l_w.rank() << ".\n");
+  EKAT_REQUIRE_MSG(
+      l_w.tags().back() == LEV,
+      "Error! The weight field must have LEV as its last dimension.\n"
+      "The weight field layout is "
+          << l_w.to_string() << ".\n");
+  EKAT_REQUIRE_MSG(l_in.rank() <= 3,
+                   "Error! The input field must be at most rank-3.\n"
+                   "The input field rank is "
+                       << l_in.rank() << ".\n");
+  EKAT_REQUIRE_MSG(l_in.rank() >= l_w.rank(),
+                   "Error! The input field must have at least as many "
+                   "dimensions as the weight field.\n"
+                   "The input field rank is "
+                       << l_in.rank() << " and the weight field rank is "
+                       << l_w.rank() << ".\n");
+  EKAT_REQUIRE_MSG(l_in.tags().back() == LEV,
+                   "Error! The input field must have a level dimension.\n"
+                   "The input field layout is "
+                       << l_in.to_string() << ".\n");
+  EKAT_REQUIRE_MSG(
+      l_in.dim(l_in.rank() - 1) == l_w.dim(l_w.rank() - 1),
+      "Error! input and weight fields must have the same dimension along "
+      "which we are taking the reducing the field (last dimensions).\n"
+      "The weight field has last dimension "
+          << l_w.dim(l_w.rank() - 1)
+          << " while "
+             "the input field has last dimension "
+          << l_in.dim(l_in.rank() - 1) << ".\n");
+  EKAT_REQUIRE_MSG(
+      l_in.dim(l_in.rank() - 1) > 0,
+      "Error! The input field must have a non-zero level dimension.\n"
+      "The input field layout is "
+          << l_in.to_string() << ".\n");
+  if(l_w.rank() == 2) {
+    EKAT_REQUIRE_MSG(
+        l_w.tags()[0] == COL && l_in.tags()[0] == COL,
```

Comment:
I'm assuming the only possible l_in types are `<COL,LEV>`, and `<COL,CMP,LEV>` (possibly with CMP repeated). If so, you could maybe consolidate this check and the next via something like
```c++
EKATE_REQUIRE_MSG (l_w.congruent(l_in.clone().strip_dim(CMP,false)),
  "Error! Incompatible layouts\n"
  "  field in: " + l_in.to_stirng() + "\n"
  "  weight: " + l_w.to_stirng() + "\n");
```
The `false` arg instruct to not throw if the dim is not found.